### ISSUE TITLE
Epsilon is applied twice?

### ISSUE THREAD

**Issue by Warry98** [TonghanWang/DOP#21]
Hello there, Thank you for sharing your code!

We are a bit confused about a possible bug in the application of epsilon during the action selection. In the controller during the forward call an epsilon floor is applied after the softmax , here's that part of the code:

````
class BasicMAC:
    ...
    def forward(self, ep_batch, t, test_mode=False):
        ...
        # Softmax the agent outputs if they're policy logits
        if self.agent_output_type == "pi_logits":
            ...
            agent_outs = th.nn.functional.softmax(agent_outs, dim=-1)

            if not test_mode:
                # Epsilon floor
                epsilon_action_num = agent_outs.size(-1)
                if getattr(self.args, "mask_before_softmax", True):
                    # With probability epsilon, we will pick an available action uniformly
                    epsilon_action_num = reshaped_avail_actions.sum(dim=1, keepdim=True).float()

                agent_outs = ((1 - self.action_selector.epsilon) * agent_outs
                               + th.ones_like(agent_outs) * self.action_selector.epsilon/epsilon_action_num)
               ...
````

but then during the multinomial action selection the actions are chosen randomly with probability epsilon as its shown in the following part of the code:

```
class MultinomialActionSelector():
    ...
    def select_action(self, agent_inputs, avail_actions, t_env, test_mode=False):
            ....
            picked_actions = Categorical(masked_policies).sample().long()

            random_numbers = th.rand_like(agent_inputs[:, :, 0])
            pick_random = (random_numbers < self.epsilon).long()
            random_actions = Categorical(avail_actions.float()).sample().long()
            picked_actions = pick_random * random_actions + (1 - pick_random) * picked_actions
            ...
            
```
In our understanding, both of this code segments are appliying the epsilon probability of picking a random action, the problem is that they're applied sequentialy, so we think the probability of picking randomly ends up being greater than intended.  Could you check if this is right?

Thank you very much!

**Comment by TonghanWang** [comment]
Thanks for pointing this out! Yes, this is a bug, and the reason for unstable learning curves. 

We are going to fix this issue. Thank you very much!

**Comment by TonghanWang** [comment]
Hi, we have removed the epsilon greedy in MultinomialActionSelector.

Actually, we were aware of this bug a few weeks ago. We tested the version with one epsilon greedy and found that the learning will be much more stable. We hope that you can find similar results :)

**Comment by Warry98** [comment]
Thank you very much for your reply!
